### PR TITLE
Updating FocusTrapZone to a function component inside React-Next

### PR DESCRIPTION
**Pull request by czearing**
#### Pull request checklist
- [X] Include a change request file using `$ yarn change`

#### Description of changes
1. Converted `FocusTrapZone` to a function component inside React-Next
2. Added perf-test tests for `FocusTrapZoneNext` and `FocusTrapZone`
3. Added vr-tests for `FocusTrapZoneNext`

#### Focus areas to test
`FocusTrapZone`

### CODE REVIEW

File: apps/perf-test/src/scenarios/FocusTrapZone.tsx [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import * as React from 'react';
```
Comment (ecraig12345) [inline apps/perf-test/src/scenarios/FocusTrapZone.tsx:1]: Do we really need a perf test for this? Each perf test does have a cost, so I'm not sure we necessarily need them for all components by default.

File: packages/react-next/src/components/FocusTrapZone/FocusTrapZone.test.tsx [hunk 2]
```diff
@@ -534,22 +543,29 @@ describe('FocusTrapZone', () => {
...
+            /*noop*/
+          },
+          stopPropagation: () => {
+            /*noop*/
+          },
+        });
       });
-      expect(lastFocusedElement).toBe(buttonB);
+
+      expect(lastFocusedElement).toBe(buttonA);
```
Comment (ecraig12345) [inline packages/react-next/src/components/FocusTrapZone/FocusTrapZone.test.tsx:568]: Why this change?

File: packages/react-next/src/components/FocusTrapZone/FocusTrapZone.test.tsx [hunk 3]
```diff
@@ -704,44 +720,41 @@ describe('FocusTrapZone', () => {
...
+          <div>
+            <FocusTrapZoneTestComponent />
+          </div>,
+        ) as unknown) as HTMLElement;
+      });
+      const buttonA = topLevelDiv!.querySelector('.a') as HTMLElement;
+      const buttonB = topLevelDiv!.querySelector('.b') as HTMLElement;
 
       expect(focusTrapZoneFocusStack.length).toBe(2);
-      const baseFocusTrapZone = focusTrapZoneFocusStack[0];
```
Comment (ecraig12345) [inline packages/react-next/src/components/FocusTrapZone/FocusTrapZone.test.tsx:729]: Did the behavior change such that this test didn't work, or was this just not necessary to cover here? (same for other changes in this test)

File: packages/react-next/src/components/FocusTrapZone/FocusTrapZone.tsx [hunk 4]
```diff
@@ -8,334 +8,389 @@ import {
...
-    }
-  }
+  const updateEventHandlers = (): void => {
+    const { isClickableOutsideFocusTrap = false, forceFocusInsideTrap = true } = props;
+    const win = getWindow(root.current)!;
 
-  private _onFocusCapture = (ev: React.FocusEvent<HTMLDivElement>) => {
-    if (this.props.onFocusCapture) {
-      this.props.onFocusCapture(ev);
+    if (forceFocusInsideTrap && !state.disposeFocusHandler) {
```
Comment (ecraig12345) [inline packages/react-next/src/components/FocusTrapZone/FocusTrapZone.tsx:246]: Is there any way we can handle click and focus handler creation and disposal with effects or something? While a 1-1 mapping from private members to internal state may technically work, it would be nice if we can take advantage of automatic updating and disposal patterns provided by hooks.

File: packages/react-next/src/components/FocusTrapZone/FocusTrapZone.tsx [hunk 5]
```diff
@@ -8,334 +8,389 @@ import {
...
+    disabled,
+  } = props;
+  const divProps = getNativeProps<React.HTMLAttributes<HTMLDivElement>>(props, divProperties);
+  const { current: state } = React.useRef<IFocusTrapZoneState>({
+    previouslyFocusedElementOutsideTrapZone: undefined,
+    previouslyFocusedElementInTrapZone: undefined,
+    disposeFocusHandler: undefined,
+    disposeClickHandler: undefined,
+    hasFocus: false,
+    prevProps: props,
```
Comment (ecraig12345) [inline packages/react-next/src/components/FocusTrapZone/FocusTrapZone.tsx:81]: Don't put this in internal state; instead use `usePrevious(props)`

File: packages/react-next/src/components/FocusTrapZone/FocusTrapZone.tsx [hunk 6]
```diff
@@ -8,334 +8,389 @@ import {
...
+};
+
+interface IFocusTrapZoneState {
+  disposeFocusHandler: (() => void) | undefined;
+  disposeClickHandler: (() => void) | undefined;
+  previouslyFocusedElementOutsideTrapZone: HTMLElement | undefined;
+  previouslyFocusedElementInTrapZone: HTMLElement | undefined;
+  hasFocus: boolean;
+  prevProps: IFocusTrapZoneProps;
+  hasPreviouslyRendered: boolean;
```
Comment (ecraig12345) [inline packages/react-next/src/components/FocusTrapZone/FocusTrapZone.tsx:54]: Remove this--looks like it's being updated but not actually read anywhere

File: packages/react-next/src/components/FocusTrapZone/FocusTrapZone.tsx [hunk 7]
```diff
@@ -8,334 +8,389 @@ import {
...
+      state.hasFocus = true;
+    },
+    [props.onFocus],
+  );
+
+  const onRootBlur = React.useCallback(
+    (ev: React.FocusEvent<HTMLDivElement>) => {
+      if (props.onBlur) {
+        props.onBlur(ev);
+      }
```
Comment (ecraig12345) [inline packages/react-next/src/components/FocusTrapZone/FocusTrapZone.tsx:154]:
```suggestion
      props.onBlur?.(ev);
```

File: packages/react-next/src/components/FocusTrapZone/FocusTrapZone.tsx [hunk 8]
```diff
@@ -8,334 +8,389 @@ import {
...
+        state.hasFocus = false;
+      }
+    },
+    [props.onBlur, elementContains, root.current],
+  );
 
-  private _onFirstBumperFocus = () => {
-    this._onBumperFocus(true);
-  };
+  const onFirstBumperFocus = useConstCallback(() => {
```
Comment (ecraig12345) [inline packages/react-next/src/components/FocusTrapZone/FocusTrapZone.tsx:171]: These can't use `useConstCallback` because `onBumperFocus` isn't constant.

**Comment by JustSlone** [comment]
@khmakoto can you help out and review this PR. Thanks!
